Add basic files for Pizza Order exercise (#2160)
exercism/javascript: exercises/concept/pizza-order/.docs/instructions.md

# Instructions

You run a pizza shop, and offer three types of pizzas:

- Margherita: $7
- Caprese: $9
- Formaggio: $10

If customers want, they can add an unlimited number of extra options: either "ExtraSauce" for $1 or "ExtraToppings" for $2.

Your task is to write code that assists the customer in figuring out the cost to them.

## Calculate the price of a pizza

Provided the pizza's name as the first argument, and an unlimited number of added options, calculate the price of the pizza in dollars.

```js
pizzaPrice('Margherita');
// => 7

pizzaPrice('Caprese', 'ExtraSauce', 'ExtraToppings');
// => 12

pizzaPrice(
  'Caprese',
  'ExtraToppings',
  'ExtraToppings',
  'ExtraToppings',
  'ExtraToppings'
);
// => 17
```

## Calculate the total price of an order

> Your function will take in a list of `PizzaOrder`s and should return the total price of the order in dollars.

Reviewer: ```suggestion
Your function is called with a list of `PizzaOrder`s and should return the total price of the order in dollars.
```

Each `PizzaOrder` has a `pizza` property - the pizza's name, and an `extras` property - the list of extra options.

```js
const margherita = new PizzaOrder('Margherita');
const caprese = new PizzaOrder('Caprese', 'ExtraToppings');
orderPrice([margherita, caprese]);
// => 18
```

> You'll realize that you can't write this using recursion, as one test with a tremendous amount of orders will raise a `Maximum call stack size exceeded`. No worries, this is intentional - try implementing this function using an imperative loop!

Reviewer: ```suggestion
You'll realize that you can't write this using recursion, as one test with a tremendous amount of orders will raise a `Maximum call stack size exceeded`.
No worries, this is intentional - try implementing this function using an imperative loop!
```

> You'll realize that you can't write this using recursion, as one test with a tremendous amount of orders will raise a `Maximum call stack size exceeded`. No worries, this is intentional - try implementing this function using an imperative loop!

Reviewer: I don't understand this:

1. The exemplar solution uses `recursion`.
2. The point of this exercise is to use recursion.
3. Why is this suggesting to use an imperative loop?

Author: So this was my confusion about this PR, as I've said in Issue #945. You say: 
> It must be solved using some imperative loop (for loop, .foreach, or best: .reduce)

I first asked a question on the issue, and then after I reread the issue, it seemed that you wanted to teach the student about these errors. I thus proceeded likewise.

> The exemplar solution uses recursion.

No, it uses `.reduce`.
```js
export function orderPrice(pizzaOrders) {
  return pizzaOrders.reduce(
    (result, order) => result + pizzaPrice(order.pizza, ...order.extras),
    0
  );
}
```

> The point of this exercise is to use recursion.

Exactly, which is why I find this to be somewhat convoluted. Should we remove the large test case, maybe? If we do that, it then doesn't teach the errors themselves (about which I agree that it's important to teach).

Reviewer: Oh right, yes I was looking at the wrong thing.

Eh. Ok, I'll change the wording a bit. Your intention (and understanding) was correct. We really do want people to hit this "issue" because it is _very_ common when trying to use recursion where you cannot, like in this case.

Thanks!

# Instructions

You run a pizza shop, and offer three types of pizzas:

- Margherita: $7
- Caprese: $9
- Formaggio: $10

If customers want, they can add an unlimited number of extra options: either "ExtraSauce" for $1 or "ExtraToppings" for $2.

Your task is to write code that assists the customer in figuring out the cost to them.

## Calculate the price of a pizza

Provided the pizza's name as the first argument, and an unlimited number of added options, calculate the price of the pizza in dollars.

```js
pizzaPrice('Margherita');
// => 7

pizzaPrice('Caprese', 'ExtraSauce', 'ExtraToppings');
// => 12

pizzaPrice(
  'Caprese',
  'ExtraToppings',
  'ExtraToppings',
  'ExtraToppings',
  'ExtraToppings'
);
// => 17
```

## Calculate the total price of an order

Your function is called with a list of `PizzaOrder`s and should return the total price of the order in dollars.
Each `PizzaOrder` has a `pizza` property - the pizza's name, and an `extras` property - the list of extra options.

```js
const margherita = new PizzaOrder('Margherita');
const caprese = new PizzaOrder('Caprese', 'ExtraToppings');
orderPrice([margherita, caprese]);
// => 18
```

> You'll realize that you can't write this using recursion, as one test with a tremendous amount of orders will raise a `Maximum call stack size exceeded`.
> No worries, this is intentional - try implementing this function using an imperative loop!

Reviewer: `````suggestion
You'll realize that you can't write this using recursion, as one test with a tremendous amount of orders will raise a `Maximum call stack size exceeded`.
No worries, this is intentional - try implementing this function using an imperative loop!
You have many options, such as, but not limited to using `reduce` or a `for` loop.

<!-- prettier-ignore-start -->
~~~~exercism/advanced
When the JavaScript interpreter is running the JavaScript code, it will keep track of which functions it has entered (started to call) on a data structure called "a stack".
When the function returns (ends), it is removed from the stack.

However, this stack has a limited size.
The most common mistake made is a recursive function that never ends.
Each call is placed on the stack, but before it returns, another call is placed on the stack.

```javascript
function kaboom() {
  kaboom()
}

kaboom()
// => RangeError: Maximum call stack size exceeded
```

The stacktrace of this error shows the same line over and over, which makes sense, because the function calls itself.
Whilst it has no real practical application in most cases, you can find out how tall that stack can get.

```javascript
let calls = 0;
function kaboom() {
  calls +=1 ;
  kaboom()
}

kaboom()
// => RangeError: Maximum call stack size exceeded

console.log(calls)
// => a number, generally higher than 10.000
```

There are only two viable solutions to a call stack error caused by a synchronous recursive function:
- ensure the functions return before the stack limit is reached, usually by adding or fixing a base case.
- rewrite the recursive function to an imperative loop, which will execute the body of the loop, without having to enter a function, thus without increasing the stack.
~~~~
<!-- prettier-ignore-end -->

`````
